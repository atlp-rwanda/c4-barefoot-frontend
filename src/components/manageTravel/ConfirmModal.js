import React from 'react';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import { Typography } from '@material-ui/core';

export default function ConfirmModal(props) {

    const {isOpen, setIsOpen, handleUpdateTravel, travelId}= props



  const handleClose = () => {
    setIsOpen({
        ...isOpen,
        open: false
    });
  };
  const handleConfirm= ()=>{
    handleClose();
    handleUpdateTravel(travelId, isOpen.action);
  }
  const handleCancel= ()=>{
    handleClose();
  }

  return (

      <Dialog
        open={isOpen.open}
        onClose={handleClose}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
      >
        <DialogTitle id="alert-dialog-title">{`Do you want to ${isOpen.action} this travel request?`}</DialogTitle>
        <DialogContent >
            <Typography 
                style={{padding: '0px 20px'}} 
                variant='subtitle1' 
                component='h6'
            >
                {`After you ${isOpen.action} this travel request, you will not be able to undo this action`}
            </Typography>

        </DialogContent>
        <DialogActions>
          <Button onClick={()=> handleCancel()} color="primary">
            Cancel
          </Button>
          <Button onClick={()=> handleConfirm()} color="primary" autoFocus>
            Confirm
          </Button>
        </DialogActions>
      </Dialog>
  );
}
