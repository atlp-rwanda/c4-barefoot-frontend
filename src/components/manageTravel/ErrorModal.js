import React, { useEffect } from 'react';
import Button from '@material-ui/core/Button';
import Snackbar from '@material-ui/core/Snackbar';
import MuiAlert from '@material-ui/lab/Alert';
import { makeStyles } from '@material-ui/core/styles';
import { Typography } from '@material-ui/core';

function Alert(props) {
  return <MuiAlert elevation={6} variant="filled" {...props} />;
}

const useStyles = makeStyles((theme) => ({
  root: {
    width: '100%',
    '& > * + *': {
      marginTop: theme.spacing(2),
    },
  },
}));

const ErrorModal= (props)=> {
  const classes = useStyles();
  // const [open, setOpen] = React.useState(false);
  const {isOpen, setIsOpen, error, clearUpdateTravelRequest}= props

  useEffect( ()=>{
    if(error){
        setIsOpen(true);
    }
  }, [])


  const handleClick = () => {
    setIsOpen(true);
  };

  const handleClose = (event, reason) => {
    if (reason === 'clickaway') {
      return;
    }

    setIsOpen(false);
    clearUpdateTravelRequest()
  };

  return (
    <div className={classes.root}>
      <Snackbar open={isOpen} autoHideDuration={6000} onClose={handleClose}>
        <Alert onClose={handleClose} severity="error">
            <Typography variant='subtitle1' component='h6' id="simple-modal-description">
               Status: { error && error.status? error.status: "Error occured"} Error: {error && error.message? error.message: (error && error.error ? error.error : "Unable to load data from server")}
            </Typography>
        </Alert>
      </Snackbar>
    </div>
  );
}
export default ErrorModal; 
