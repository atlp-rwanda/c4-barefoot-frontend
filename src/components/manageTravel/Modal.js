import React, {useEffect} from 'react'
import {connect} from 'react-redux'
import { makeStyles, Dialog, DialogActions, DialogContent, DialogTitle, Slide, Grid, Paper, Card, CardActions, CardActionArea, Avatar, CardMedia, Typography, Button, Divider } from '@material-ui/core';
import CloseIcon from '@material-ui/icons/Close';
import { getSingleTravelRequest } from "../../redux/actions/singleTravelAction";
import { Skeleton } from '@material-ui/lab';

const useStyle = makeStyles(theme => ({
    dialog:{
        padding: 60,
    },
    grid:{
        width:'100%',
        justifyContent:'center'
    },
    originCityImage:{
        width:'100%',
        height:'400px'
    },
     gridDestination:{
        width:'300px',
        justifyContent:'center'
    },
   destinationCityImage:{
        width:'100%',
        height:'200px'
    },
    closeIcon:{
        position:'absolute',
        justifyContent:'space-between',
        color:'red',
        top: '2rem',
        right:'10px',
        '&:hover':{
            cursor: 'pointer'
        }
    },
    travelAddress:{
        display:'flex',
        justifyContent:'space-between',
        marginTop:'15px',
        marginBottom:'15px',
        '&:hover':{
            borderBottom:'2px solid lightgrey',
            padding:'10px',
            boxShadow:true
        }
    },
    addressPaper:{
        border:'none',
        boxShadow:'0px 0px 0px 0px',
        marginTop:'20px',
        padding:"5px"
    },
    travelDetails:{
        justifyContent:'center',
        alignItems:'center',
        boxShadow:'none'
    },
    gridDestination:{
        flexGrow:1
    },
    buttons:{
    float:'inline-end',
    marginRight:'20px',
    justifyContent:'space-evenly',
    display:'inline-block',
    backgroundColor: 'yellow'
  },
  approveButton:{
      backgroundColor:'green'
  },
  rejectButton:{
    //   marginRight:'20px'
  },
  travelData:{
      width:'100%',
      justifyContent:'center',
      alignItems:'center',
      display:"flex"
  },
  profile:{
      display:'flex',
      justifyContent:'center',
      alignItems:'center',
      marginTop:'10px'
  },
  status:{
      display:'flex',
      justifyContent:'center',
      alignItems:'center',
      marginTop:'10px'
  }, 
  trips: {
      textAlign:'center',
      justifyContent:'center',
      alignItems: 'center'
  }
}))

function Modal(props) {
    const classes = useStyle()
    const {title, children, openModal, onClose, setOpenModal, singleTravel} = props;

    useEffect(() => {
        getSingleTravelRequest()
        
    }, [])
    const travelrequest = singleTravel.travel[0]
    // const requester = singleTravel.travel[0].requesterInfo

    console.log(travelrequest);

    const cardImage = "https://images.pexels.com/photos/67636/rose-blue-flower-rose-blooms-67636.jpeg";
    const Transition = React.forwardRef(function Transition(props, ref) {
  return <Slide direction="up" ref={ref} {...props} />;
});
    return (
       <Dialog open={openModal} TransitionComponent={Transition} onClose={onClose} maxWidth='md' className={classes.dialog} >
           { singleTravel.travel.length === 0 ? (
           <div>
               <Skeleton variant='text'/>
               <Skeleton variant='circle' height={40} width={40} />
               <Skeleton variant='rect' width={500} height={300}/>
           </div>
           )
           :
           <React.Fragment>
            <DialogTitle style={{textAlign:'center'}}>
                Travel request details
                <CloseIcon className={classes.closeIcon} />
            </DialogTitle>
            <DialogContent>
                <Card>
                <CardActionArea>
                    <Grid className={classes.grid}>
                        
                        <Paper className={classes.originCityImage}>
                            <img src={cardImage} alt='card image' className={classes.originCityImage}/>
                        </Paper>
                       
                    </Grid>
                    <Grid className={classes.travelData}>
                        <Paper className={classes.travelDetails}>
                            <Typography variant='h6' component='h6' >Travel information</Typography>
                            <div className={classes.profile}> 
                                <Avatar alt="Remy Sharp" src={"requester.profile_picture"} /> 
                                <Typography style={{marginLeft:'20px'}}>{"requester.first_name"} {"requester.last_name"}</Typography>
                            </div>
                            <div className={classes.status}>
                                <Typography variant='h6' component='h6'>status:</Typography>
                                <Typography style={{marginLeft:'20px'}}>{travelrequest.status}</Typography>
                            </div>
                        </Paper>
                    </Grid>
                    <Divider />
                    <Typography varient='h6' component="h6" className={classes.trips} style={{textAlign:'center', marginTop:'20px'}}>{travelrequest.Trip.length} Trips</Typography>
                    {travelrequest.Trip.map( data => (
                    <Grid className={classes.travelAddress} key={data.tripId}>
                    <Paper className={classes.addressPaper}>
                        <Typography variant='h6' component='h6'>Lacation destination</Typography>
                        <Typography>{data.originCity} To {data.destination} </Typography>
                        {/* <Typography>Kampala, Uganda To Kenya, Nairobi </Typography>
                        <Typography>Kenya, Nairobi To Egypt, Cairo </Typography>
                        <Typography>{travelrequest.status} </Typography> */}
                    </Paper>
                     <Paper className={classes.addressPaper}>
                        <Typography variant='h6' component='h6'> Travel Date</Typography>
                        <Typography>08/10/2020</Typography>
                    </Paper>
                     <Paper className={classes.addressPaper}>
                        <Typography variant='h6' component='h6'>Reason</Typography>
                        <Typography>{data.reason}</Typography>
                    </Paper>
                     <Paper className={classes.addressPaper}>
                        <Typography variant='h6' component='h6'>Date of Returning</Typography>
                        <Typography>01/05/2021</Typography>
                    </Paper>
                    </Grid>

                    ))}
                    
                    {/* <Grid className={classes.gridDestination}>
                        <Grid>
                            <div className={classes.buttons}>
                                <Button variant="contained" color="primary" className={classes.approveButton}>Approve</Button>
                                <Button variant="contained" color="secondary" className={classes.rejectButton}>Reject</Button>
                            </div>
                        </Grid>
                    
                     <Grid></Grid>
                    </Grid> */}

                </CardActionArea>
                </Card>
            </DialogContent>
            </React.Fragment>}

       </Dialog>
    )
}

const mapStateToProps = state => ({
    singleTravel:state.manageSingleTravel
})
// export default Modal
export default connect(mapStateToProps, {getSingleTravelRequest})(Modal)
