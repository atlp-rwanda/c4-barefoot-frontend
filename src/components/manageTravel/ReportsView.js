import React, { useState, useEffect } from "react";
import {
  Typography,
  makeStyles,
  useTheme,
  Drawer,
  Grid,
  TextField,
  Card,
  GridList,
  GridListTile,
  CardActionArea,
  CardMedia,
  CardContent,
  IconButton,
  CardActions,
  Avatar,
  Button,
  Paper,
  Modal,
  Container,
  Box,
} from "@material-ui/core";
import { connect } from "react-redux";
import Model from "./viewTravelModal";
import { Skeleton } from "@material-ui/lab";
import Pagination from "@material-ui/lab/Pagination";
import { getTravelRequest } from "../../redux/actions/fetchTravelRequestAction";
import {
  updateSingleTravelRequest,
  clearUpdateTravelRequest,
} from "../../redux/actions/updateTravelRequestAction";
import {
  clearSingleRequest,
  getSingleTravelRequest,
} from "../../redux/actions/singleTravelAction";
import AccountCircleIcon from "@material-ui/icons/AccountCircle";
import ErrorModal from "./ErrorModal";
import TravelRequestCard from "./TravelRequestCard";
import SuccessModal from "./SuccessModal";

const useStyles = makeStyles((theme) => ({
  container: {
    minHeight: "78vh",
    display: "flex",
    alignItems: "center",
    padding: "0%",
    marginTop: "20px",
    flexDirection: "column",
    paddingBottom: "20px",
  },
  cardContainer: {
    display: "flex",
    flexDirection: "column",
    minHeight: "75vh",
    position: "relative",
    justifyContent: "space-between",
  },
  root: {
    padding: "20px",
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "center",
    boxShadow: "none",
    marginTop: "20px",
    position: "relative",
  },
  details: {
    display: "flex",
    flexDirection: "column",
    justifyContent: "flex-start",
  },
  content: {
    flex: "1 0 auto",
  },

  image: {
    width: "100%",
    height: "300px",
  },
  name: {
    marginLeft: "10px",
    marginTop: "-5px",
  },
  controls: {
    display: "flex",
    alignItems: "center",
    paddingLeft: theme.spacing(1),
    paddingBottom: theme.spacing(1),
    justifyContent: "space-between",
  },

  requesterProfi: {
    display: "flex",
    width: "350px",
    flexDirection: "row",
    justifyContent: "flex-start",
  },
  imageContainer: {
    width: "300px",
    height: "200px",
    marginLeft: "0%",
    overflow: "hidden",
  },

  paganete: {
    marginTop: "20px",
    width: "100%",
    display: "flex",
    justifyContent: "center",
    marginBottom: "15px",
    "& > *": {
      marginTop: theme.spacing(2),
    },
  },
  buttons: {
    float: "inline-end",
    marginRight: "20px",
    justifyContent: "space-evenly",
    display: "inline-block",
  },
}));

const filter = (travels, category) => {
  let filteredArray = [];
  switch (category) {
    case "approved":
      filteredArray = travels.filter((trav) => {
        return trav.travelRequestInfo.status === category;
      });
      return filteredArray;
    case "pending":
      filteredArray = travels.filter((trav) => {
        return trav.travelRequestInfo.status === category;
      });
      return filteredArray;

    case "canceled":
    case "rejected":
      filteredArray = travels.filter((trav) => {
        return (
          trav.travelRequestInfo.status === "canceled" ||
          trav.travelRequestInfo.status === "rejected"
        );
      });
      return filteredArray;

    case "done":
      filteredArray = travels.filter((trav) => {
        if (trav.travelRequestInfo.Trip.length !== 0) {
          return (
            trav.travelRequestInfo.Trip[trav.travelRequestInfo.Trip.length - 1]
              .returnDate < Date.now() &&
            trav.travelRequestInfo.Trip[trav.travelRequestInfo.Trip.length - 1]
              .returnDate
          );
        }
      });
      return filteredArray;
    default:
      return [];
  }
};

function ReportsView(props) {
  const classes = useStyles();
  const theme = useTheme();
  const loading = props.travelRequest.loading;
  const travel = props.travelRequest.travel;
  const error = props.travelRequest.error;
  const [openModal, setOpenModal] = useState(false);
  const [isErrorModalOpen, setIsErrorModalOpen] = useState(false);
  const { updateSingleTravel } = props;
  const success = updateSingleTravel.travel;
  const [isSuccessModalOpen, setIsSuccessModalOpen] = useState(false);

  const { clearUpdateTravelRequest } = props;

  const category = props.category;
  const [modalUsage, setModalUsage] = useState("view");
  const [page, setPage] = useState(1);
  const page_size = 5;

  const filtered = filter(travel, category);

  const handlePage = (e, value) => {
    setPage(value);
  };

  useEffect(() => {
    props.getTravelRequest();
  }, []);

  useEffect(() => {
    if (success) {
      handleModalClose();
    }
  }, [success]);

  //updatting travel request
  const handleUpdateTravel = (id, action) => {
    props.updateSingleTravelRequest(id, action);
  };
  //formating dsate

  // openning modal
  const handleModalOpen = () => {
    setOpenModal(true);
  };
  //getting single travel
  const handleSingleTravel = (id, usage) => {
    handleModalOpen();
    props.getSingleTravelRequest(id);
    setModalUsage(usage);
  };

  // closing modal
  const handleModalClose = () => {
    setOpenModal(false);
    props.clearSingleRequest();
    //  props.clearUpdateTravelRequest();
    props.getTravelRequest();
  };

  const paginate = (array, page_number) => {
    // human-readable page numbers usually start with 1, so we reduce 1 in the first argument
    return array.slice((page_number - 1) * page_size, page_number * page_size);
  };
  const paginated = paginate(filtered, page);

  return (
    <div className={classes.container} style={{ boxShadow: "none" }}>
      <Typography variant="h6" component="h6">
        Direct report travel Request
      </Typography>
      {loading ? (
        <>
          <Card className={classes.root} style={{ boxShadow: "none" }}>
            <Paper className={classes.imageContainer}>
              <Skeleton
                variant="rect"
                animation="wave"
                height="300px"
                width="320px"
              />
            </Paper>
            <div className={classes.details}>
              <CardContent className={classes.content}>
                <div className={classes.requesterProfi}>
                  <Skeleton variant="circle" width={40} height={40} />
                  <Skeleton
                    variant="text"
                    width="190px"
                    height="20"
                    className={classes.name}
                  />
                  <Skeleton
                    variant="text"
                    height="20"
                    width="100px"
                    style={{ float: "right" }}
                  />
                </div>
                <Skeleton
                  variant="text"
                  width="300px"
                  height="40px"
                  style={{ marginLeft: "30px" }}
                />
                <Skeleton
                  variant="text"
                  width="100px"
                  height="40px"
                  style={{ marginLeft: "30px" }}
                />
                <div className={classes.controls}>
                  <Skeleton
                    variant="text"
                    width="100px"
                    height="40px"
                    style={{ marginLeft: "23px" }}
                  />
                  <div className={classes.buttons}>
                    <Skeleton variant="text" width="150px" height="40px" />
                  </div>
                </div>
              </CardContent>
            </div>
          </Card>
        </>
      ) : (
        <Container maxWidth="md" className={classes.cardContainer}>
          <Box>
            {filtered.length !== 0 ? (
              paginated.map((trav) => (
                <Card className={classes.root} key={trav.travelId}>
                  <TravelRequestCard
                    travel={trav.travelRequestInfo}
                    handleSingleTravel={handleSingleTravel}
                    category={category}
                    userInfo={trav.userInfo}
                    accomodationsInfo={trav.accommodationInfo}
                  />

                  <Model
                    openModal={openModal}
                    setOpenModal={setOpenModal}
                    onClose={handleModalClose}
                    handleUpdateTravel={handleUpdateTravel}
                    updateSingleTravel={updateSingleTravel}
                    usage={modalUsage}
                    accomodationsInfo={trav.accommodationInfo}
                  >
                    <div>this is details of travel request</div>
                  </Model>
                </Card>
              ))
            ) : (
              <center>
                <Box style={{ paddingTop: "50px" }}>
                  <Typography variant="subtitle1" component="h6">
                    No {category} travel request found
                  </Typography>
                </Box>
              </center>
            )}
          </Box>
          <Box>
            <ErrorModal
              isOpen={isErrorModalOpen}
              setIsOpen={setIsErrorModalOpen}
              error={error}
            />
            <div className={classes.paganete}>
              {filtered.length !== 0 && filtered.length / page_size > 1 && (
                <Pagination
                  count={Math.ceil(filtered.length / page_size)}
                  color="primary"
                  onChange={handlePage}
                />
              )}
            </div>

            {success && (
              <SuccessModal
                isOpen={isSuccessModalOpen}
                setIsOpen={setIsSuccessModalOpen}
                success={success}
                clearUpdateTravelRequest={clearUpdateTravelRequest}
                handleModalClose={handleModalClose}
              />
            )}
          </Box>
        </Container>
      )}
    </div>
  );
}
const mapStateToProps = (state) => ({
  travelRequest: state.manageTravel,
  singleTravel: state.manageSingleTravel,
  updateSingleTravel: state.updateTravel,
});

export { ReportsView };
export default connect(mapStateToProps, {
  getTravelRequest,
  getSingleTravelRequest,
  updateSingleTravelRequest,
  clearSingleRequest,
  clearUpdateTravelRequest,
})(ReportsView);
