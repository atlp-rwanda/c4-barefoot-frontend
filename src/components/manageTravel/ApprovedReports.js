import React from 'react'
import ReportsView from './ReportsView';


function manageTravelDashboard() {
    
    return (
        <ReportsView category={"approved"} />
    )
}

export default manageTravelDashboard;
