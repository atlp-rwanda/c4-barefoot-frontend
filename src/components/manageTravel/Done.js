import React from 'react'
import ReportsView from './ReportsView';


function Done() {
    
    return (
        <ReportsView category={"done"} />
    )
}

export default Done;
