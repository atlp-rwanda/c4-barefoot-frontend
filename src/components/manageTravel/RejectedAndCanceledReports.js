import React from 'react'
import ReportsView from './ReportsView';


function RejectedAndCanceled() {
    
    return (
        <ReportsView category={"rejected"} />
    )
}

export default RejectedAndCanceled;
