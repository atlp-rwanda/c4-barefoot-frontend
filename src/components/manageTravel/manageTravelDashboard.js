import React from 'react'

import ReportsView from './ReportsView';


function manageTravelDashboard(props) {

    return (
        <ReportsView category={'pending'} />
    )
}

export default manageTravelDashboard;
