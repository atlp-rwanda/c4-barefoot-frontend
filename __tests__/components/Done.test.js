import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Done } from '../../src/components/manageTravel/Done';


describe(' <Done />', ()=>{
    let wrapper;
    it( 'It should render the Done component', ()=>{
        wrapper= shallow(<Done />)
        const component= toJson(wrapper);
        expect(component).toMatchSnapshot();
    })


})

