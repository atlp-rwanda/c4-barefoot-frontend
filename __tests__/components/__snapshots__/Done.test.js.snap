// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` <Done /> It should render the Done component 1`] = `
<Connect(ReportsView)
  category="done"
/>
`;
